app/models/coupon.rb

```diff
@@ -129,18 +129,53 @@ def to_csv
   def can_edit?
     self.redemptions.empty?
   end
+
+  def validate_new_offers(offer_codes)
+    offer_codes = offer_codes.select(&:present?)
+    if offer_codes.blank?
+      self.errors.add(:offer, "code is required")
+    end
+    offer_codes.map do |offer_code|
+      offer = Offer.new(code: offer_code)
+      if !offer.valid?
+        self.errors.add(:offer, "code #{offer_code} is not valid")
+      end
+      offer
+    end
+  end
+
+  def update_offers(offer_codes)
+    updated_codes =  offer_codes.select(&:present?)
+
+    if updated_codes.blank?
+      self.errors.add(:offer, "code is required")
+      return []
+    end
+
+    existing_codes = self.offers.pluck(:code)
+    added_codes = updated_codes - existing_codes
+    removed_codes = existing_codes - updated_codes
+
+    removed_codes.each do |code|
+      Offer.find_by(offerable_id: self.id, code: code).destroy
+    end
+    added_codes.each do |code|
+      self.offers << Offer.new(code: code)
+    end
+  end
    
   private
   
   # find the coupon, or raise an exception if that coupon is not valid
-  def self.find_coupon(coupon_code, user_id = nil)
+  def self.find_coupon(coupon_code, offer_code, user_id = nil)
     coupon = Coupon.with_code(coupon_code.upcase).first
     raise CouponNotFound if coupon.nil?
     if user_id && coupon.redemptions.find_by_user_id(user_id)
       raise CouponAlreadyRedeemedByUser
     end
     raise CouponRanOut if coupon.redemptions_count >= coupon.how_many
     raise CouponExpired if coupon.expiration < Time.now.to_date
+    raise CouponNotApplicable if coupon.offers.present? && !coupon.offers.pluck(:code).include?(offer_code)
```

Comment:
In case we'd like to change it -- this is the line that allows generic coupons to work until they run out / expire.

Reply:
Awesome!